File: 938_Range_Sum_of_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


# Time Complexity : O(n) [n = number of nodes in Tree]
# Space Complexity : O(n) [n = number of nodes in Tree]
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
#
# Problem approach
# 1. Traverse the Tree in preorer way
# 2. Add the value of current node if it falls within the range
# 3. Recusively call the dfs on left and right child if it falls within the range
class Solution:
    def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
        res = 0
        def dfs(root):
            nonlocal res
            ## base
            if not root:
                return
            
            ## body
            if root.val >= L and root.val <= R:
                res+=root.val
            
            if root.val > L:
                dfs(root.left)
            
            if root.val < R:
                dfs(root.right)
        
        dfs(root)
        return res

File: 938-RangeSumOfBST.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        // return bfs(root,low,high);
        return dfs(root,low,high);
    }
    
    
    int bfs(TreeNode* root, int low, int high) {
        queue<TreeNode*> q;
        q.push(root);
        int sum=0;
        while(!q.empty()) {
            auto node = q.front();q.pop();
            if(node->val >= low and node->val <= high) {
                sum += node->val;
            }
            
            if(node->left) q.push(node->left);
            if(node->right) q.push(node->right);
        }
        return sum;
    }
    
    int dfs(TreeNode* root, int low, int high) {
        stack<TreeNode*> st;
        st.push(root);
        int ans = 0;
        while(!st.empty()) {
            auto node = st.top();st.pop();
             if(node->val >= low and node->val <= high) {
                ans += node->val;
            }
            
            if(node->left) st.push(node->left);
            if(node->right) st.push(node->right);
        }
        
        return ans;
    }
    
};